apps/designer/app/shared/router-utils/user-content-params.ts

```diff
@@ -0,0 +1,90 @@
+export type CanvasRouteMode = "edit" | "preview" | "published";
+const modes = ["edit", "preview", "published"] as CanvasRouteMode[];
+const getMode = (url: URL): CanvasRouteMode => {
+  const modeParam = url.searchParams.get("mode");
+  const mode =
+    modeParam === null ? "published" : modes.find((mode) => mode === modeParam);
+  if (mode === undefined) {
+    throw new Error(`Invalid mode "${modeParam}"`);
+  }
+  return mode;
+};
+
+// A subtype of Request. To make testing easier.
+type MinimalRequest = {
+  url: string;
+  headers: { get: (name: string) => string | null };
+};
+
+const getRequestHost = (request: MinimalRequest): string =>
+  request.headers.get("x-forwarded-host") || request.headers.get("host") || "";
+
+export const getUserContentBaseUrl = (request: MinimalRequest): string => {
+  const { USER_CONTENT_BASE_URL } = process.env;
+  if (USER_CONTENT_BASE_URL !== undefined && USER_CONTENT_BASE_URL !== "") {
+    return USER_CONTENT_BASE_URL;
+  }
+
+  // Local development special case
+  const host = getRequestHost(request);
+  if (
+    process.env.NODE_ENV === "development" &&
+    /^(.*\.)?localhost(:\d+)?$/.test(host)
+  ) {
+    return `http://${host.split(".").pop()}`;
+  }
+
+  // Vercel preview special case
+  if (
+    (process.env.VERCEL_ENV === "preview" ||
+      process.env.VERCEL_ENV === "development") &&
+    typeof process.env.VERCEL_URL === "string"
+  ) {
+    return `https://${process.env.VERCEL_URL}`;
+  }
+
+  throw new Error("Could not determine user content host");
+};
+
+export const getUserContentParams = (
+  request: MinimalRequest
+):
+  | {
+      projectId: string;
+      mode: CanvasRouteMode;
+      pathname: string;
+    }
+  | {
+      projectDomain: string;
+      mode: CanvasRouteMode;
+      pathname: string;
+    }
+  | undefined => {
+  const url = new URL(request.url);
+
+  const requestHost = getRequestHost(request);
+  const userContentHost = new URL(getUserContentBaseUrl(request)).host;
+
+  if (process.env.USER_CONTENT_REQUIRE_SUBDOMAIN !== "true") {
```

Comment:
please check that all new env variables are validated if they are required in env-check.ts

Reply:
Done